File: haskell/bob/src/Bob.hs
module Bob (responseFor) where

import Data.Char (isAlpha, isSpace, isUpper)
import Data.List (isSuffixOf)

responseFor :: String -> String
responseFor xs
  | silence = "Fine. Be that way!"
  | yellingAQuestion = "Calm down, I know what I'm doing!"
  | question = "Sure."
  | shouting = "Whoa, chill out!"
  | otherwise = "Whatever."
  where
    text = filter (not . isSpace) xs
    alpha = filter isAlpha text

    silence = null text
    yellingAQuestion = shouting && question
    question = "?" `isSuffixOf` text 
    shouting = (not . null $ alpha) && all isUpper alpha

File: typescript/bob/bob.ts
export const hey = (message: string): string => {
  const containsLetter = /[a-zA-Z]/g.test(message)
  const yelling = containsLetter && message === message.toUpperCase()
  const asking = message.trimEnd().at(-1) === '?'
  const silence = !message.replace(/\s/g, '').length
  
  if (silence) return "Fine. Be that way!"
  if (yelling && asking) return "Calm down, I know what I'm doing!"
  if (yelling) return 'Whoa, chill out!'
  if (asking) return 'Sure.'

  return 'Whatever.'
}
